docs/adrs/0079-api-versioning.md

```diff
@@ -0,0 +1,124 @@
+
+## API Versioning Strategy
+
+### SemVer
+The Prime API follows the SemVer (Semantic Versioning) convention, using a three-part version number: Major.Minor.Patch. Here's what each part signifies:
+
+Major versions incorporate significant changes, including backward-incompatible API changes.
+
+Minor versions add new features in a backwards-compatible manner.
+
+Patch versions include backwards-compatible bug fixes.
+
+For example, a version could be noted as "2.1.3," where '2' is the major version, '1' is the minor version, and '3' is the patch version.
+
+### Options for Versioning Strategy
+There are a few options for implementing api versioning in our prime API.
+
+* Option 1: Include version in the URL patterns.
+  * Pros: This is the simplest option. We currently have `/v1/` in the paths of all of our APIs, and the implementation of our routing initialization can be easily modified to include multiple paths. If we went with some other option we would want to remove the `v1` from our outward facing APIs.
+  * Cons: Could lead to a big footprint in the codebase when branching the API.
+* Option 2: Indicate version based on query parameters.
+  * Pros: Easy to default to the latest version
+  * Cons: Difficult to route requests. Would need to modify our current routing processes to account for this. Would have a similar effect on the codebase footprint as option 1.
+* Option 3: Custom headers for versioning.
+  * Pros: Similar to the previous 2 but does not clutter up the URL path with version information
+  * Cons: Would require us to create the infrastructure to route based on these custom headers. Would have a similar effect on the codebase footprint as option 1.
+* Option 4: Versioning through Content Negotiation or media type. Customize the accept media type header to specify the versioning.
+  * Pros: Can version directly at the resource level.
+  * Cons: Difficult to test. Would be a heavy lift as we would have to refactor our routing infrastructure. Potentially confusing for the consumers of our API.
+
+**Chosen Alternative:** Option 1. This would be the most straightforward approach and easiest lift as it
+ builds off of our current routing for our API.
+
+### Codebase management
+In order to mitigate the potential for our codebase to swell in size as we branch off versions of our
+APIs, we need to address which endpoints will get pulled into the new versions.
+
+* Option 1: Pull in each endpoint as the need to implement breaking changes arises.
+  * When the first endpoint needs to be refactored to introduce a breaking change to the prime API, then a version 2 API will be created containing that one endpoint. That endpoint would be deprecated over time in version 1. More endpoints would be brought in to version 2 as the need arises.
+  * Pros: Much easier lift. Lessens the size of our codebase.
+  * Cons: Could be more confusing for the consumers of our API since not all endpoints are on each version
```

Comment:
I have thought some about it. I think it could definitely be beneficial and would be nice to have. I am not sure about whether it would simplify our testing or not. 

However, I have thought about some possible points of confusion. Take the example where we introduce a version 2 api with a few endpoints, and then route all other endpoints to the v1 api. The prime makes the switch to version 2, after they have made the changes to their system to handle those breaking changes. Later, when we introduce a new endpoint to version 2, it will effectively be introducing a breaking change to our v2 api without incrementing the major version. While it won't technically be doing that on our end, we will just be adding a new endpoint to the v2 api, to the prime it will appear that way. They will have switched to version 2, we will then tell them to switch back to v1 for a certain endpoint because the v2 endpoint now has breaking changes. Granted, this could definitely be mitigated by clear communication.

Reply:
Ooh that's a good counterpoint scenario that I wasn't thinking of.  I might wonder if really that v1 endpoint that didn't get bumped to v2 with the others, shouldn't go straight to v3 in terms of releases of the API.